tests/test_2408_layoutbuilder_in_numba.py

```diff
@@ -0,0 +1,842 @@
+    subbuilder.append(1.1)
+    assert len(builder) == 1
+    array = builder.snapshot()
+    assert ak.to_list(array) == [1.1]
+
+    builder.append_null()
+    subbuilder.append(np.nan)
+    assert len(builder) == 2
+
+    data = np.array([3.3, 4.4, 5.5], dtype=np.float64)
+
+    builder.extend_valid(3)
+    subbuilder.extend(data)
+    assert len(builder) == 5
+
+    builder.extend_null(2)
+    subbuilder.append(np.nan)
+    subbuilder.append(np.nan)
+
+    assert len(builder) == 7
+
+    builder.append_valid()
+    subbuilder.append(8)
+    assert len(builder) == 8
+
+    builder.append_valid()
+    subbuilder.append(9)
+    assert len(builder) == 9
+
+    builder.append_valid()
+    subbuilder.append(10)
+    assert len(builder) == 10
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [1.1, None, 3.3, 4.4, 5.5, None, None, 8, 9, 10]
+
+
+def test_Union_Numpy_ListOffset():
+    builder = lb.Union(
+        np.int64,
+        [
+            lb.Numpy(np.float64),
+            lb.ListOffset(np.int64, lb.Numpy(np.int32)),
+        ],
+    )
+    assert len(builder) == 0
+
+    # error = ""
+    # assert builder.is_valid(error) == True
+
+    one = builder.append_index(0)
+    one.append(1.1)
+
+    # assert builder.is_valid(error) == True
+
+    two = builder.append_index(1)
+    list = two.begin_list()
+    list.append(1)
+    list.append(2)
+    two.end_list()
+
+    # assert builder.is_valid(error) == True
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [1.1, [1, 2]]
+
+
+def test_Union_ListOffset_Record():
+    builder = lb.Union(
+        np.int64,
+        [
+            lb.ListOffset(np.int64, lb.Numpy(np.int32)),
+            lb.Record([lb.Numpy(np.float64), lb.Numpy(np.int64)], ["x", "y"]),
+        ],
+    )
+    assert len(builder) == 0
+
+    one = builder.append_index(0)
+    list = one.begin_list()
+    list.append(1)
+    list.append(3)
+    one.end_list()
+
+    two = builder.append_index(1)
+    x = two.field("x")
+    y = two.field("y")
+
+    x.append(1.1)
+    y.append(11)
+
+    builder.append_index(0)
+    list = one.begin_list()
+    list.append(5.5)
+    one.end_list()
+
+    builder.append_index(1)
+    x.append(2.2)
+    y.append(22)
+
+    array = builder.snapshot()
+    assert ak.to_list(array) == [[1, 3], {"x": 1.1, "y": 11}, [5], {"x": 2.2, "y": 22}]
+
+
+def test_unbox():
+    @numba.njit(debug=True)
+    def f1(x):
+        x  # noqa: B018 (we want to test the unboxing)
+        return 3.14
+
+    builder = lb.Numpy(np.int32)
+    f1(builder)
+
+    builder = lb.Empty()
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.Numpy(np.float64))
+    f1(builder)
+
+    builder = lb.ListOffset(np.int32, lb.Empty())
+    f1(builder)
+
+    builder = lb.List(np.int8, lb.Numpy(np.int64))
+    f1(builder)
+
+    builder = lb.List(np.int32, lb.Empty())
+    f1(builder)
+
+    # FIXME:
+    builder = lb.ListOffset(np.int32, lb.ListOffset(np.int64, lb.Numpy(np.int64)))
```

Comment:
@jpivarski - I need some help. I wonder if I need to use `numba.deferred_type` here. Currently, the first `ListOffset` type is correct, but the content of the second one is a `bound method NumpyType.type`:
```
src/awkward/_connect/numba/layoutbuilder.py:581: in __init__
    super().__init__(dmm, fe_type, members)
        __class__  = <class 'awkward._connect.numba.layoutbuilder.ListOffsetModel'>
        dmm        = <numba.core.datamodel.manager.DataModelManager object at 0x7fd318b6d040>
        fe_type    = ak.numba.lb.ListOffset(int32, ak.numba.lb.ListOffset(int64, ak.numba.lb.Numpy(int64)))
        members    = [('offsets', ak.GrowableBuffer(int32)),
 ('content',
  ak.numba.lb.ListOffset(int64, <bound method NumpyType.type of <class 'awkward._connect.numba.layoutbuilder.NumpyType'>>))]
        self       = <awkward._connect.numba.layoutbuilder.ListOffsetModel object at 0x7fd3197dd3a0>

```

Reply:
I've never used `deferred_type` before (I had to look it up) and I doubt it's needed here. You'd have to defer a type for circular references, but all layout builders are strictly tree-like.

Can you say something more about what the issue is? I don't have any context.

Could it be that you want the return value of a `type` method/property and not the function-as-a-value? Is `NumpyType` the Numba type of a `lb.Numpy` instance?